blockchain/state_transition.go

```diff
@@ -357,8 +359,14 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 
 	// Defer transferring Tx fee when DeferredTxFee is true
 	if st.evm.ChainConfig().Governance == nil || !st.evm.ChainConfig().Governance.DeferredTxFee() {
-		effectiveGasPrice := msg.EffectiveGasPrice(nil)
-		st.state.AddBalance(st.evm.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveGasPrice))
+		if rules.IsMagma {
+			effectiveGasPrice := st.gasPrice
+			txFee := getBurnAmountMagma(new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveGasPrice))
+			st.state.AddBalance(st.evm.Rewardbase, txFee)
+		} else {
+			effectiveGasPrice := msg.EffectiveGasPrice(nil)
+			st.state.AddBalance(st.evm.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveGasPrice))
```

Comment:
yes. The logic of providing rewards to proposers instead of rewardbase before the magma hardfork is intentional. As mentioned in solution 2, several service chains currently in operation are using the configuration that causes this issue. Therefore, modifying this logic could potentially lead to other problems (e.g., bad blocks). Consequently, the fix for this issue has been postponed until after the magma hardfork.

Reply:
Ah, I understand now that an updated binary and a non-updated binary may have different ledgers, necessitating a hardfork to incorporate the changes. I learned your careful thought process. Then, I'm curious about why you decided to change it to a block proposer from the initially given [`rewardbase`](https://github.com/klaytn/klaytn/pull/1886/files/68fbe63036aee84375ece2dd1bd29455ac4291f6#r1270282767)?

Reply:
This is a bit complex logic, so I'll give an example along with an explanation.

For example, if deferredtxfee=false, policy=simple
CalcDeferredRewardSimple is used in two cases
**1.** in Block Finalize ([Link](https://github.com/JayChoi1736/klaytn/blob/deferfix/consensus/istanbul/backend/engine.go#L500C3-L500C3))
  - this logic calculates rewardMap which includes block minting amount and total txfee. 
  - the rewardMap is used in DistributeBlockReward and it increases corresponding accounts' balance
  - because deferredtxfee=false, total tx fee already added during ApplyTransaction. So totaltxfee = 0 and it will not affect on the rewardMap.
  - (But before Magma hardfork, totaltxfee wil not be set as 0)

**2.** in getRewards API ([Link](https://github.com/JayChoi1736/klaytn/blob/deferfix/reward/reward_distributor.go#L201))
  - this logic calculates getRewards API result
  - but CalDeferredRewardSimple will not shows exact result in deferredTxfee=false condition, because it sets totaltxfee=0 due to **1.** 
  - so we changes the rewardSpec in post-processing as [Link](https://github.com/JayChoi1736/klaytn/blob/deferfix/reward/reward_distributor.go#L215C16-L237)
  - (But before Magma hardfork, totaltxfee also added in proposer, so we have to put those logic in post-processing)


the rewardBase what you mentioned is in the post-processing. if rules.IsMagma=false, we have to show exact Rewardmap which totaltxfee goes to both rewardbase and proposer.

Reply:
Thanks for the detailed explanation.

The question was, that I was curious why the APIs revised the reward list (specifically reward receiver, the change `proposer` to `rewardaddress`) to be returned. After reviewing again, I got the simple answer. I'll appreciate it if you verify it.

In core logic:
Before-Magma: Total tx fee should be added to `coinbase` address. (i.e., total tx fee -> `coinbase`)
After-Magma: Burn the tx fee with halving, and give the residual to the `rewardbase` addresses. (i.e., 2/totla tx fee -> `rewardbase`)

Thus, the API should also align with what the core logic does. Then, the change `proposer` to  `rewardaddress` does make sense.
Before-Magma: The reward receiver should be `coinbase`, which is the proposer.
After-Magma: The reward receiver should be `rewardbase`.

Reply:
yes, you are right.

Reply:
Thank you